### PR TITLE
Fix two improper casts in HavingSpecMetricComparator.

### PR DESCRIPTION
**Pull request by gianm**
Fixes two things:

1. An improper double-to-long cast when comparing double metrics to any
   kind of value, which was a regression from #4883.
2. An improper double-to-long cast when comparing a long/int metric to a
   double/float value: the value was cast to long/int, drawing strange
   conclusions like int 100 matching a havingSpec of equalTo(100.5).

Marked 0.12.0 since one of these was introduced in #4883, new in 0.12.0.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java [hunk 1]
```diff
@@ -46,18 +47,22 @@ static int compare(Row row, String aggregationName, Number value, Map<String, Ag
...
-      if (metricValueObj instanceof Long) {
-        long l = ((Long) metricValueObj).longValue();
-        return Longs.compare(l, value.longValue());
+      if (metricValueObj instanceof Long || metricValueObj instanceof Integer) {
+        long n = ((Number) metricValueObj).longValue();
+
+        if (value instanceof Long || value instanceof Integer) {
+          return Longs.compare(n, value.longValue());
+        } else if (value instanceof Double || value instanceof Float) {
+          return BigDecimal.valueOf(n).compareTo(BigDecimal.valueOf(value.doubleValue()));
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: please add comment about reason to not use Doubles.compare and floats.compare ?

**Comment by gianm** [comment]
@nishantmonu51 added comments; thanks for review.

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java [hunk 2]
```diff
@@ -46,18 +47,26 @@ static int compare(Row row, String aggregationName, Number value, Map<String, Ag
...
+          // represent all float values.
+          return BigDecimal.valueOf(n).compareTo(BigDecimal.valueOf(value.doubleValue()));
+        } else {
+          throw new ISE("Number was[%s]?!?", value.getClass().getName());
+        }
       } else if (metricValueObj instanceof Float) {
-        float l = ((Float) metricValueObj).floatValue();
-        return Floats.compare(l, value.floatValue());
+        float n = (Float) metricValueObj;
+        return Floats.compare(n, value.floatValue());
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: can we combine float and double also similar to long and int... also check type of value instance and use BigDecimal if it is not double? ... to keep it similar to long/int case.

Comment (gianm) [reply processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: @himanshug I edited the code to simplify it, and factored out the BigDecimal stuff to a helper method. Now the branches are more similar.

Comment (himanshug) [reply processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: thanks

**Pull request merged**